### DISCUSSION TITLE
Renovate does not always correctly update gradle dependency verification data

### DISCUSSION THREAD

**Discussion by connyduck** [renovatebot/renovate#29784] [Request Help]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

When updating the Android Gradle plugin dependencies, Renovate fails to correctly update the dependency verification data. See [here](https://github.com/tuskyapp/Tusky/pull/4512) for an example.
Is there some configuration I'm missing or is this a bug?
It works fine with other dependencies.

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by secustor** [comment]
What are you referring here to with `Renovate fails to correctly update the dependency verification data`?

If you mean the notification, this is intended behavior as you have pushed a commit on top of the PR.

**Reply by connyduck** [reply]
Renovate wants to update com.android.application to v8.5.0. But the PR it sends only partially updates the `gradle/verification-metadata.xml`, so the build fails.

```
Dependency verification failed for configuration ':app:detachedConfiguration2'
  One artifact failed verification: aapt2-8.5.0-11315950.pom (com.android.tools.build:aapt2:8.5.0-11315950) from repository Google
```

I then have to manually add the missing data (the commit I pushed on top) to get the project to build again.
I would expect renovate to send a correct PR by itself and save my time.

**Comment by rarkins** [comment]
What Gradle commands did renovate run when it created the PR? Which additional commands did you run to create your commit?

**Reply by connyduck** [reply]
> What Gradle commands did renovate run when it created the PR?

No idea, where do I see this and can I configure that somehow?

> Which additional commands did you run to create your commit?

`./gradlew --write-verification-metadata sha256 app:lintGreenDebug app:assembleGreenDebug`

& then manually add the data for aapt2, which has different artifacts for windows, linux and macos.

**Reply by rarkins** [reply]
You can get that from the debug logs when the commit was made. If that's too old then you can tick the rebase/retry box in the PR to force override the old commits (including your own).

Your commands seem very specific to your project though. How would expect Renovate to learn/guess which commands to run?

**Reply by ShadowRZ** [reply]
> You can get that from the debug logs when the commit was made. If that's too old then you can tick the rebase/retry box in the PR to force override the old commits (including your own).

In my case it uses `:dependencies` task, since https://github.com/renovatebot/renovate/pull/29602

> Your commands seem very specific to your project though. How would expect Renovate to learn/guess which commands to run?

Actually I think there's some other issue, in my case after the PR was created by Renovate, verification metadata changes only contained the artifact and Gradle module metadata file, however running other common tasks (for example `:app:assembleDebug` for build debug APKs or `testDebugUnitTests` to run unit tests) somehow make Gradle wanted to verify additional POM files, even with `<verify-metadata>false</verify-metadata>`

**Reply by jamietanna** [reply]
Given a couple of projects I've worked with that use Gradle Verification Metadata, the `dependencies` task has ended up finding all relevant configurations - not discounting your point-of-view, only mentioning it as a fact it can be based on Gradle config, plugins used, etc

That might be a question more for the Gradle project - to help work out whether there's anything else Renovate needs to invoke - before we can make any changes here

**Reply by ShadowRZ** [reply]
In my case, the AboutLibraries Android Gradle plugin was resolving these POM files since it needs to grab artifact descriptions and licenses which are only present in POM files. I ended up updating the generated metadata files manually.

> That might be a question more for the Gradle project - to help work out whether there's anything else Renovate needs to invoke - before we can make any changes here

I think so in general, especially when the issue involves detached configurations: in the original author's case, when running `:app:assmebleDebug`, AGP resolves aapt2 with a detached configuration (log file says `Dependency verification failed for configuration ':app:detachedConfiguration2'`, note the `detachedConfiguration2`), and these detached configurations can't be auto discovered by Gradle when running `:dependencies` task. I also noted some other cases where Renovate's Gradle updates failed on CI because verification metadata doesn't contain artifacts resolved from detached configurations, which are only resolved when you run some production task (like `assembleDebug`, `lint`)

**Comment by connyduck** [comment]
> How would expect Renovate to learn/guess which commands to run?

🤷  It could check our CI config or have a configuration option to specify the gradle command.
I guess this is more of a feature request than a bug report then.

**Comment by kailash-everlaw** [comment]
I recently ran into this issue - in my scenario I think the root cause was that Gradle's `--write-verification-metadata` behaves differently depending on the current state of Gradle's dependency cache [1]. If you are using Gradle lockfiles, Renovate will first update the lockfile with a `gradlew` command and then run a second command to update the dependency checksums. The first command will cause the dependency cache to be populated, which causes the second command to not always update all checksums.

One possible fix could be to use `--refresh-dependencies` when updating verification metadata, since that instructs Gradle to ignore the dependency cache. Setting `GRADLE_USER_HOME` to an empty directory also might fix the issue.

[1] https://github.com/gradle/gradle/issues/19228
